lib/spack/spack/solver/asp.py

```diff
@@ -1017,8 +1017,10 @@ def pkg_rules(self, pkg, tests):
                 values = [variant.default]
 
             for value in sorted(values):
-                self.gen.fact(fn.variant_possible_value(pkg.name, name, value))
-                if hasattr(value, "when"):
+                if getattr(value, "when", True) is not True:  # when=True means unconditional
+                    if value.when is False:
+                        continue
+
```

Comment:
Just a suggestion to improve readability, and avoid nested ifs:
```suggestion
                # when=True means unconditional
                value_condition = getattr(value, "when", True)

                if value_condition is False:
                    continue

                if isinstance(value_condition, str):  
```